torch/csrc/DataLoader.cpp

```diff
@@ -0,0 +1,253 @@
+#include <sys/wait.h>
+#include <set>
+#include <atomic>
+#include <signal.h>
+#include "THP.h"
+
+// In cases like DataLoader, if a worker process die due to bus error/segfault
+// or just hang, the main process, if implemented with
+// multiprocessing.queue.SimpleQueue, will hang waiting for data. This is
+// difficult to avoid on PyTorch side as it can be caused by limited shm, or
+// other libraries users call in the workers. The following methods is an effort
+// to do our best provide some error message to users when such unfortunate
+// events happen.
+
+// TODO: The following don't work on Windows. Specifically, waitpid calls and
+// SIGCHLD handler.
+
+#ifndef _WIN32
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig, siginfo_t *info, void *ctx)                 \
+{                                                                             \
+    write(STDERR_FILENO, ERROR_MSG, sizeof(ERROR_MSG) / sizeof(char));        \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+// signal(2) is really not portable. So use sigaction.
+// http://man7.org/linux/man-pages/man2/signal.2.html
+static int setSignalHandler(int signal, void(*handler)(int, siginfo_t *, void *), struct sigaction *old_sa_ptr)
+{
+  struct sigaction sa;
+  sa.sa_sigaction = handler;
+  sa.sa_flags = SA_RESTART|SA_SIGINFO|SA_NOCLDSTOP;
+  sigemptyset(&sa.sa_mask);
+  return sigaction(signal, &sa, old_sa_ptr);
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
+  HANDLE_TH_ERRORS
+  int error = 0;
+  error |= setSignalHandler(SIGBUS, &handler_SIGBUS, NULL) != 0;
+  error |= setSignalHandler(SIGSEGV, &handler_SIGSEGV, NULL) != 0;
+  return PyBool_FromLong(!error);
+  END_HANDLE_TH_ERRORS
+}
+
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static std::atomic<pid_t *> worker_pids;
+static std::atomic<size_t> num_worker_pids(0);
+// Pipe used as a lock to avoid update of the above and SIGCHLD handler in parallel.
+static int comm_pipe[2] = {-1, -1};
+
+static int updatePIDsArray() {
+
+  size_t new_size =  worker_pid_vec.size();
+  auto new_ptr = (pid_t *)malloc(sizeof(pid_t) * new_size);
+  for (size_t idx = 0; idx < new_size; idx++) {
+    new_ptr[idx] = worker_pid_vec[idx];
+  }
+
+  // Block SIGCHLD handler for this thread so SIGCHLD handler can't interrupt
```

Comment:
Okay, here is what happens. When we update the array, the updating thread:

1. blocks SIGCHLD signal from this thread, so the handler can't interrupt the update on this thread.
2. reads from a pipe, which acts like a lock, and writes to it after finishing. (The handler similarly also reads from the lock before reading the contents of the array, and write afterwards. )

Since multiple threads can't run this function in parallel due to GIL, I think this should work.